# Assignment

In this assignment we want to get comfortable with loading and manipulating data in Python. While future assignments will focus more using structured data which we can load into a `DataFrame` using `pandas`, this assignment is focused on semi-structured data and how we can "flatten" it and then load it into other formats. The objective is to see how data flows in Python from one object to another and what advantages and disadvantages each offers.

Let's read the `books.json` data set and display the first item in it.

In [12]:
import json
with open('../data/books.json', encoding = 'utf-8') as f: # f = file
    books_dict = json.load(f)

from pprint import pprint
pprint(books_dict[0]) # print information for the first book

{'_id': 1,
 'authors': ['W. Frank Ableson', 'Charlie Collins', 'Robi Sen'],
 'categories': ['Open Source', 'Mobile'],
 'isbn': '1933988673',
 'longDescription': 'Android is an open source mobile phone platform based on '
                    'the Linux operating system and developed by the Open '
                    'Handset Alliance, a consortium of over 30 hardware, '
                    'software and telecom companies that focus on open '
                    'standards for mobile devices. Led by search giant, '
                    'Google, Android is designed to deliver a better and more '
                    'open and cost effective mobile experience.    Unlocking '
                    "Android: A Developer's Guide provides concise, hands-on "
                    'instruction for the Android operating system and '
                    'development tools. This book teaches important '
                    'architectural concepts in a straightforward writing style '
                    

In [13]:
print(len(books_dict))

431


1. Write a program that goes through the entire data and extracts the following information:  <span style="color:red" float:right>[4 point]</span>

  - title of the book
  - name of the first author
  - name of the second author (if book has more than one author)
  - number of authors
  - ISBN
  - if the word "data" is in the book's description
  - the number of words in the book's description
  - the year the book was published

  Of course because JSON data doesn't necessarily enforce any sort of schema, we can't be sure that the information we are trying to extract exists for every book. For example, if the book only has one author, then there is no second author. So use `try` and `except` as you loop through every book and skip to the next item every time some information is missing.

  Store the extracted data in a list named `rows` whose elements are tuples, one tuple per book. For example, the first element of `rows` stores the tuple for the first book and should look like this:

        ('Unlocking Android', 'W. Frank Ableson', 'Charlie Collins', 3, '1933988673', True, 252, 2009)

In [14]:
from dateutil.parser import parse # For year of publishing

# Empty list to hold tuples
rows = [] 

# Loop through each book in dictionary
for book in books_dict:
    
    # Attempt to extract a title. If no title found, default value is none.
    try:
        title = book['title']
    except:
        title = None
        
    # Attempt to extract first author(0). If first author not found, default value is none.    
    try:
        author_1 = book['authors'][0]
    except:
        author_1 = None
        
    # Attempt to count number of authors. If none found, default value is 0.
    try:
        num_authors = len(book['authors'])
    except:
        num_authors = 0
        
    # If there is more than one author, extract the name of the the second author.
    if num_authors > 1:
        author_2 = book['authors'][1]
    else:
        author_2 = None
    
    # Attempt to extract the ISBN. If not found, default value is none.
    try:
        isbn = book['isbn']
    except:
        isbn = None
        
    # I made an assumption that the word data is in the long description.
    # Attempt to find data in long description. If not found, default value is False.
    try:    
        has_data = True if 'data' in book['longDescription'] else False
    except:
        has_data = False
    
    # I made the assumption that the word is in the long description.
    # Attempt to count the number of words in the description of the book.
    # If an error occurs, default values is none.
    try:
        words = [''.join(c for c in word if c.isaplpha() or c=="'") for word in description[longDescrition].split()]
    # Removing special characters from list of words. If a single character word is not a special character, keep it.   
        desc_len = len(words)
    except:
        desc_len = None
    
    # Attempt to extract the year the book was published. If not found, the default value is none.
    try:
        year_published = parse(book['publishedDate']['$date']).year
    except:
        year_published = None
    
    # Creating a tuple for the book.
    cur_tuple = tuple([title, author_1, author_2, num_authors, isbn, has_data, desc_len, year_published])
    
    # Append the current book's tuple to rows list created earlier.
    rows.append(cur_tuple)

print(len(rows))
pprint(rows)

431
[('Unlocking Android',
  'W. Frank Ableson',
  'Charlie Collins',
  3,
  '1933988673',
  True,
  None,
  2009),
 ('Android in Action, Second Edition',
  'W. Frank Ableson',
  'Robi Sen',
  2,
  '1935182722',
  False,
  None,
  2011),
 ('Specification by Example',
  'Gojko Adzic',
  None,
  1,
  '1617290084',
  False,
  None,
  2011),
 ('Flex 3 in Action',
  'Tariq Ahmed with Jon Hirschi',
  'Faisal Abid',
  2,
  '1933988746',
  True,
  None,
  2009),
 ('Flex 4 in Action',
  'Tariq Ahmed',
  'Dan Orlando',
  4,
  '1935182420',
  True,
  None,
  2010),
 ('Collective Intelligence in Action',
  'Satnam Alag',
  None,
  1,
  '1933988312',
  True,
  None,
  2008),
 ('Zend Framework in Action',
  'Rob Allen',
  'Nick Lo',
  3,
  '1933988320',
  True,
  None,
  2008),
 ('Flex on Java',
  'Bernerd Allmon',
  'Jeremy Anderson',
  2,
  '1933988797',
  True,
  None,
  2010),
 ('Griffon in Action',
  'Andres Almiray',
  'Danno Ferrin',
  4,
  '1935182234',
  False,
  None,
  2012),
 ('OSGi in D

1. Reason. I wrote a program to extract specific book data (title of book, name of author, etc). To do this I created a list called rows, then specified details of each book data to information specifications using the try/except method, and ended by creating a tuple with the book data, appending it to the list named rows.

1. Conclusion. Utilized pretty print to acquire the information of the desired extracted information.

2. Save the content of `rows` in a SQL-like table using `sqlite3`, and choose the appropriate column types. <span style="color:red" float:right>[2 point]</span> 

  As your column names use the following:

  - `title`
  - `author_1`
  - `author_2`
  - `num_authors`
  - `isbn`
  - `has_data`
  - `desc_len`
  - `year_published`

In [15]:
import sqlite3

# Creating a connection to an in-memory SQLite database as shown in class.
connection = sqlite3.connect(':memory:') 
cursor = connection.cursor()

cursor.execute('''CREATE TABLE books
             (title text, author_1 text, author_2 text, num_authors int, isbn text, has_data text, desc_len text, year_published int)''')

# Insert the rows list into the new table
cursor.executemany('INSERT INTO books VALUES (?,?,?,?,?,?,?,?)', rows)

# Save the changes
connection.commit()

2. Reason. Utilzed sqlite 3 to save the content of the rows with the given column types.

2. Conclusion. Inserted the rows list into the new table with specified columns and saved the changes.

3. Write a SQL query against the table to show all books that (1) contain the word "data" and (2) have more than 3 authors. Store the result of the query in an object called `books_table`, then close the connection. <span style="color:red" float:right>[2 point]</span>

In [16]:
# Query books with the word "data" in the description and more than 3 authors.
books_table = cursor.execute('SELECT * FROM books WHERE has_data = 1 AND num_authors > 3').fetchall()
print(books_table)

# Close connection to the database.
connection.close()

[('Flex 4 in Action', 'Tariq Ahmed', 'Dan Orlando', 4, '1935182420', '1', None, 2010), ('Spring Dynamic Modules in Action', 'Arnaud Cogoluegnes', 'Thierry Templier', 4, '1935182307', '1', None, 2010), ('Android in Practice', 'Charlie Collins', 'Michael D. Galpin', 4, '1935182927', '1', None, 2011), ('Portlets and Apache Portals', 'Stefan Hepper', 'Peter Fischer', 5, None, '1', None, 2005), ('Struts in Action', 'Ted N. Husted', 'Cedric Dumoulin', 4, '1932394249', '1', None, 2002), ('Taming Text', 'Grant S. Ingersoll', 'Thomas S. Morton', 4, '193398838X', '1', None, 2012), ('Groovy in Action', 'Dierk Koenig with Andrew Glover', 'Paul King', 4, '1932394842', '1', None, 2007), ('Groovy in Action, Second Edition', 'Dierk König', 'Guillaume Laforge', 8, '1935182447', '1', None, None), ('NHibernate in Action', 'Pierre Henri Kuate', 'Tobin Harris', 5, '1932394923', '1', None, 2009), ('Tuscany SCA in Action', 'Simon Laws', 'Mark Combellack', 5, '1933988894', '1', None, 2011), ('Entity Framework

3. Reason. Created an SQL query against the table containing the word "data" and over three authors. 

3. Conclusion. Stored the query result in books_table and closed the connection to the database.

SQL tables are not the only way, and definitely not the most straightforward way to store and manipulate data in Python. A format that's more popular with data scientist is to use the `pandas` library to create a `DataFrame`. This library has a lot of functionality that makes it easy to run the common tasks data scientists do with data.

4. Read the data from the above query into a `DataFrame` and call it `books_df`. HINT: Use `pd.DataFrame` and specify meaningful column names to use for the columns. <span style="color:red" float:right>[1 point]</span>

In [17]:
import pandas as pd

# Create a list of column names for the dataframe
columns = ['title', 'author_1', 'author_2', 'num_authors', 'isbn', 'has_data', 'desc_len', 'year_published']

# Load the books_table subject into a dataframe with the above column names
books_df = pd.DataFrame(data = books_table, columns = columns)

# Change the data type of the has_data column back to boolean
books_df['has_data'] = books_df['has_data'].astype('bool')

# Change the data type of the year published column to an integer vice float as default
books_df['year_published'] = books_df['year_published'].astype('Int64')

books_df

title  \
0                   Flex 4 in Action   
1   Spring Dynamic Modules in Action   
2                Android in Practice   
3        Portlets and Apache Portals   
4                   Struts in Action   
5                        Taming Text   
6                   Groovy in Action   
7   Groovy in Action, Second Edition   
8               NHibernate in Action   
9              Tuscany SCA in Action   
10      Entity Framework 4 in Action   
11         SQL Server MVP Deep Dives   
12         SQL Server MVP Deep Dives   
13           ASP.NET MVC 2 in Action   
14           ASP.NET MVC 4 in Action   
15   JUnit in Action, Second Edition   
16                 Node.js in Action   
17                   HTML5 in Action   

                                             author_1                author_2  \
0                                         Tariq Ahmed             Dan Orlando   
1                                  Arnaud Cogoluegnes        Thierry Templier   
2                                     Charlie Collins       Michael D. Galpin   
3                                       Stefan Hepper           Peter Fischer   
4                                       Ted N. Husted         Cedric Dumoulin   
5                                  Grant S. Ingersoll        Thomas S. Morton   
6                     Dierk Koenig with Andrew Glover               Paul King   
7                                         Dierk König       Guillaume Laforge   
8                                  Pierre Henri Kuate            Tobin Harris   
9                                          Simon Laws         Mark Combellack   
10                                   Stefano Mostarda        Marco De Sanctis   
11              Contributions from 53 SQL Server MVPs  Edited by Paul Nielsen   
12  Contributions from 53 SQL Server MVPs; Edited ...           Kalen Delaney   
13                                    Jeffrey Palermo           Ben Scheirman   
14                                    Jeffrey Palermo            Jimmy Bogard   
15                                     Petar Tahchiev             Felipe Leme   
16                                      Mike Cantelon             Marc Harter   
17                                       Rob Crowther              Joe Lennon   

    num_authors           isbn  has_data desc_len  year_published  
0             4     1935182420      True     None            2010  
1             4     1935182307      True     None            2010  
2             4     1935182927      True     None            2011  
3             5           None      True     None            2005  
4             4     1932394249      True     None            2002  
5             4     193398838X      True     None            2012  
6             4     1932394842      True     None            2007  
7             8     1935182447      True     None            <NA>  
8             5     1932394923      True     None            2009  
9             5     1933988894      True     None            2011  
10            4     1935182188      True     None            2011  
11            8     1935182048      True     None            2009  
12            7  9781935182047      True     None            2009  
13            6     193518279X      True     None            2010  
14            6     1617290416      True     None            2012  
15            5     1935182021      True     None            2010  
16            5     1617290572      True     None            2013  
17            5     1617290491      True     None            2014

4. Reason. Read the data from the SQL query into a dataframe using pd.dataframe and named it books_df with specific column names for the columns.

4. Conclusion. Created a dataframe named books_df and displayed it showing columns with specific information.

5. Display the first few columns of a `DataFrame` by calling its `head` method. <span style="color:red" float:right>[1 point]</span>

In [52]:
# Displaying the first five rows of the dataframe using .head
books_df.head()

title            author_1           author_2  \
0                  Flex 4 in Action         Tariq Ahmed        Dan Orlando   
1  Spring Dynamic Modules in Action  Arnaud Cogoluegnes   Thierry Templier   
2               Android in Practice     Charlie Collins  Michael D. Galpin   
3       Portlets and Apache Portals       Stefan Hepper      Peter Fischer   
4                  Struts in Action       Ted N. Husted    Cedric Dumoulin   

   num_authors        isbn  has_data desc_len  year_published  
0            4  1935182420      True     None            2010  
1            4  1935182307      True     None            2010  
2            4  1935182927      True     None            2011  
3            5        None      True     None            2005  
4            4  1932394249      True     None            2002

In [54]:
books_df.dtypes

title             object
author_1          object
author_2          object
num_authors        int64
isbn              object
has_data            bool
desc_len          object
year_published     Int64
dtype: object

5. Reason. To display the first five rows of the dataframe named books_df using the .head method.

5. Conclusion. Displayed the first five rows of the datagrame named books_df.

Remember how earlier we said that a `DataFrame` is built on top of `numpy` arrays? Another way of saying it is that a `DataFrame` is an **abstraction** on top of `numpy` arrays: i.e. a `DataFrame` is a more **high-level** object than a `numpy` array. 

6. Call the `values` attribute of your `DataFrame` to convert it into a numpy array and display the first 3 elements of the array. <span style="color:red" float:right>[1 point]</span>

In [55]:
import numpy as np

# Create a numpy array from the dataframe's values method.
np_array = books_df.values

# Display the first three elements of the array.
print(np_array[:3])

[['Flex 4 in Action' 'Tariq Ahmed' 'Dan Orlando' 4 '1935182420' True None
  2010]
 ['Spring Dynamic Modules in Action' 'Arnaud Cogoluegnes'
  'Thierry Templier' 4 '1935182307' True None 2010]
 ['Android in Practice' 'Charlie Collins' 'Michael D. Galpin' 4
  '1935182927' True None 2011]]


6. Reason. To call the values attribute of the books_df dataframe and convert it into a numpy array. Then display the first three elements of that array.

6. Conclusion. Created a numpy array from the dataframe's values method and displayed the first three elements of the array using the print function. Displayed is the first three rows of books_df. 

Now you can judge which object is more "user-friendly". That's one of the things that abstractions allow us to do: build more user-friendly (abstract) objects from less user-friendly (but more fundamental) objects.

# End of assignment